metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+    assert(node > 0 && node <= num_nodes());
+    std::string kmer = get_node_sequence(node);
+    auto call_impl = [&](const auto &dict) {
+        using kmer_t = typename std::remove_reference<decltype(dict)>::type::kmer_t;
+        sshash::neighbourhood<kmer_t> nb = dict.kmer_forward_neighbours(kmer.c_str(), true);
+        for (size_t i = 0; i < kmer_t::alphabet_size; i++) {
+            if (nb.forward[i].kmer_id != sshash::constants::invalid_uint64) {
+                callback(sshash_to_graph_index(nb.forward[i].kmer_id), kmer_t::alphabet[i],
+                         nb.forward[i].kmer_orientation);
+            }
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        call_impl(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        call_impl(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        call_impl(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+void DBGSSHash::call_incoming_kmers_with_rc(
+        node_index node,
+        const std::function<void(node_index, char, bool)>& callback) const {
+    assert(node > 0 && node <= num_nodes());
+    std::string kmer = get_node_sequence(node);
+    auto call_impl = [&](const auto &dict) {
+        using kmer_t = typename std::remove_reference<decltype(dict)>::type::kmer_t;
+        sshash::neighbourhood<kmer_t> nb = dict.kmer_backward_neighbours(kmer.c_str(), true);
+        for (size_t i = 0; i < kmer_t::alphabet_size; i++) {
+            if (nb.backward[i].kmer_id != sshash::constants::invalid_uint64) {
+                callback(sshash_to_graph_index(nb.backward[i].kmer_id), kmer_t::alphabet[i],
+                         nb.backward[i].kmer_orientation);
+            }
+        }
+    };
+
+    if (auto dict64 = dynamic_cast<Dictionary<Kmer64>*>(dict_.get())) {
+        call_impl(*dict64);
+    } else if (auto dict128 = dynamic_cast<Dictionary<Kmer128>*>(dict_.get())) {
+        call_impl(*dict128);
+    } else if (auto dict256 = dynamic_cast<Dictionary<Kmer256>*>(dict_.get())) {
+        call_impl(*dict256);
+    } else {
+        common::logger->error("");
+        throw std::runtime_error("k fail");
+    }
+}
+
+size_t DBGSSHash::outdegree(node_index node) const {
+    size_t res = 0;
+    adjacent_outgoing_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+size_t DBGSSHash::indegree(node_index node) const {
+    size_t res = 0;
+    adjacent_incoming_nodes(node, [&](node_index) { ++res; });
+    return res;
+}
+
+void DBGSSHash::call_nodes(
+        const std::function<void(node_index)>& callback,
+        const std::function<bool()> &terminate) const {
+    for (size_t node_idx = 1; !terminate() && node_idx <= dict_->size(); ++node_idx) {
+        callback(node_idx);
+        if (mode_ == CANONICAL && !terminate()) {
```

Comment:
Does `call_nodes` provide any guarantees on the order of calls? It may be better to run rc nodes in a second loop, so that we don't constantly jump between the two node types in memory.

Reply:
in CANONICAL mode, only one orientation is stored in memory. The index of the reverse complement is simply the same index offset by the size of the underlying dictionary, so both orientations of a k-mer, in fact, point to the same "location" in the underlying memory.

If we were to do these separately instead of interleaved, we'd essentially have to do two passes through the dictionary instead of one.

Reply:
Isn't doing one pass and doing something twice within it essentially the same as doing two passes and doing something once within them? But at least it'd guarantee that indices are called in an increasing order, so there should be no jumps in memory between regular and reversed nodes, if they're accessed by callback?

Reply:
hmm...yea, I just realized that if some external process relies on node indices to access an array, then this will indeed lead to jumps in memory. So let's do this as two loops then, since it's a cheap operation anyway.